feat(RELEASE-1179): docs for single component mode (#135)
konflux-ci/docs: docs/modules/ROOT/pages/advanced-how-tos/testing-releasing-single-component.adoc

= Testing and Releasing a Single Component

By default, {ProductName} treats applications as a set of dependent components. A example would be frontend and backend
components. A user may typically want to build, test, and release these components together
to ensure there are compatible.

Applications may also consist of independent components. If a user wants to group components together in an application
a failure in one component will normally prevent any components from releasing.
In **Single Component Mode**, however, each component in an application is separated so that one component's failing
Enterprise Contract Policy will not prevent the other components from releasing.

== Enabling Single Component Mode

Single component mode works by analyzing the Snapshot that is being used for Integration Tests and Releasing.
A Snapshot created due to a single component being built is the only Snapshot type that may be used for this
mode.

=== Enabling Single Component Mode for Enterprise Contract Integration Test Scenarios

Single Component Mode can be enabled either by making a change to the Integration Test Scenario custom resource

> or by using the HAC UI.

Reviewer: ```suggestion
or by using the UI.
```

Author: fixed

= Testing and Releasing a Single Component

By default, {ProductName} treats applications as a set of dependent components. A example would be frontend and backend
components. Typically, a user would want to build these components, while testing and releasing them together
to ensure there are compatible.

However, applications can also be viewed simply as a group of non-dependent components.

Consider this:

* You have an application that contains 5 components.
* You need to release a single component to fix a bug that a customer is experiencing.
* While attempting to release, another component has a new violation therefore causing the Enterprise Contract
policy validation to fail.
** This violation may be due to a rule that used to be a warning but has expired and is now a violation.

Wouldn't it be nice to be able to ignore that failing component and simply release the component you are targeting?

> Enter **Single Component Mode**!

Author: I am easy. fixed

= Testing and Releasing a Single Component

> By default, Konflux treats applications as a set of dependent components. A example would frontend and backend

Reviewer: ```suggestion
By default, {ProductName} treats applications as a set of dependent components. A example would be frontend and backend
```

Author: fixed

> components. Typically, a team would want to build these components, while testing and releasing them together

Reviewer: ```suggestion
components. Typically, a user would want to build these components, while testing and releasing them together
```

Author: fixed

policy validation to fail.
** This violation may be due to a rule that used to be a warning but has expired and is now a violation.

Wouldn't it be nice to be able to ignore that failing component and simply release the component you are targeting?

Enter **Single Component Mode**!

== Enabling Single Component Mode

Single component mode works by analyzing the Snapshot that is being used for Integration Tests and Releasing.
A Snapshot created due to a single component being built is the only Snapshot type that may be used for this
mode.

=== Enabling Single Component Mode for Enterprise Contract Integration Test Scenarios

Single Component Mode can be enabled either by making a change to the Integration Test Scenario custom resource
or by using the HAC UI.

* Update the custom resource by adding the parameter *SINGLE_COMPONENT* with a value of *true*

[source,yaml]
.integration-test-scenario.yaml
----
apiVersion: appstudio.redhat.com/v1beta2
kind: IntegrationTestScenario
metadata:
  name: grouping-enterprise-contract-single-component
  namespace: shebert-tenant
spec:
  application: grouping
  contexts:
    - description: Application testing
      name: application

>   params:

Reviewer: Is this the only part that is new? If so, we can just use a snippet like we do for defining build customizations so that they don’t need to figure out what information they don’t need to include.

Author: fixed snippet

    - name: SINGLE_COMPONENT
      value: 'true'
  resolverRef:
    params:
      - name: url
        value: 'https://github.com/konflux-ci/build-definitions'
      - name: revision
        value: main
      - name: pathInRepo
        value: pipelines/enterprise-contract.yaml
    resolver: git
----

[NOTE]
====
You need to apply the change using *kubectl*
====

* Alternatively, you may edit the Integration Test Scenario using the HAC UI as follows:

image::single-component-integration-test-scenario.png[role="border" alt="Editing your Integration Test Scenario"]

=== Enabling Single Component Mode for Releases

Single Component Mode can be enabled for Releases by making a change to the ReleasePlanAdmission custom resource.

> * Update the custom resource by adding the field *singleComponentMode* under _.spec.data_ with a value of *true*

Reviewer: You can also add the annotation to the RPA CR page with a description and link the two pages instead.

[source,yaml]
.release-plan-admission.yaml
----
apiVersion: appstudio.redhat.com/v1alpha1
kind: ReleasePlanAdmission
metadata:
  name: single-component-rpa
  namespace: managed-release-team-tenant
  labels:
    release.appstudio.openshift.io/auto-release: 'true'
spec:
  applications:
    - grouping
  data:
    mapping:
      components:
        - name: comp1
          repository: quay.io/scoheb/comp1-release
          tags:
            - latest
        - name: comp2
          repository: quay.io/scoheb/comp2-release
          tags:
            - latest
    singleComponentMode: true
  origin: shebert-tenant
  pipeline:
    pipelineRef:
      params:
        - name: url
          value: 'https://github.com/konflux-ci/release-service-catalog.git'
        - name: revision
          value: production
        - name: pathInRepo
          value: pipelines/push-to-external-registry/push-to-external-registry.yaml
      resolver: git
    serviceAccountName: release-service-account
    timeouts:
      pipeline: 4h0m0s
  policy: e2e-single-component-policy
----

[NOTE]
====
You need to apply the change using *kubectl*
====

> === Analyzing Test Results and Releases in Single Component Mode

Reviewer: Is this analyzing or just verifying? Since we go over ITS and releases in subcategories, we can probably just call this verifying generically.

Author: fixed

* Once your changes have been applied, you can verify that the mode is in fact activated.

> ==== Verifying Enterprise Contract Integration Test Scenario results

Reviewer: Is this only for the EC ITS or ITS generally?

Author: EC ITS. I updated the next line to clarify

* Navigate to the logs of your Integration Test Scenario pipeline.
* Locate the *reduce* step in the log.
* It should resemble the following output:

[source]
--

> [reduce] Single Component mode? true

Reviewer: For this and the next snippet, is the content relevant or just the fact that the reduce parts ran? It might be simpler to parse if we remove the unnecessary details specific to your run.

Author: removed some of the snippet
